From e31a6e41506e41e11b8717164fc1e90fdb7c4d5b Mon Sep 17 00:00:00 2001
From: soulteary <soulteary@gmail.com>
Date: Fri, 7 May 2021 11:54:06 +0800
Subject: [PATCH] Optimize the size of the built image

---
 .dockerignore |  7 +++++++
 Dockerfile    | 18 +++++++++++-------
 2 files changed, 18 insertions(+), 7 deletions(-)
 create mode 100644 .dockerignore

diff --git a/.dockerignore b/.dockerignore
new file mode 100644
index 0000000..070f67a
--- /dev/null
+++ b/.dockerignore
@@ -0,0 +1,7 @@
+.git*
+docs
+*.md
+Dockerfile
+LICENSE
+modd.conf
+.dockerignore
\ No newline at end of file
diff --git a/Dockerfile b/Dockerfile
index a6580b7..e100832 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,13 +1,17 @@
-# build stage
-FROM golang:alpine AS builder
-RUN apk add --no-cache build-base
+FROM golang:1.16.4-alpine AS builder
 WORKDIR /src
+RUN adduser --disabled-password --uid 10001 shioriuser
+RUN apk --update add ca-certificates musl-dev gcc
+COPY go.mod go.sum ./
+RUN go env -w GO111MODULE=on && go env -w  GOPROXY=https://goproxy.io,direct
 COPY . .
-RUN go build
+RUN go generate ./... && CGO_ENABLED=1 go build -a -ldflags '-linkmode external -extldflags "-static" -s -w' .
 
-# server image
-FROM golang:alpine
+FROM scratch
+COPY --from=builder /etc/passwd /etc/passwd
 COPY --from=builder /src/shiori /usr/local/bin/
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+USER shioriuser
 ENV SHIORI_DIR /srv/shiori/
 EXPOSE 8080
-CMD ["/usr/local/bin/shiori", "serve"]
+CMD ["/usr/local/bin/shiori", "serve"]
\ No newline at end of file
