From d5eaeed0a38c163be7f623b1bf0e73043022873f Mon Sep 17 00:00:00 2001
From: soulteary <soulteary@gmail.com>
Date: Fri, 7 May 2021 12:21:58 +0800
Subject: [PATCH] Use environment variables to set the default user

---
 internal/webserver/handler-api.go | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/internal/webserver/handler-api.go b/internal/webserver/handler-api.go
index bea32dd..82ba1a3 100644
--- a/internal/webserver/handler-api.go
+++ b/internal/webserver/handler-api.go
@@ -74,9 +74,19 @@ func (h *handler) apiLogin(w http.ResponseWriter, r *http.Request, ps httprouter
 	accounts, err := h.DB.GetAccounts(searchOptions)
 	checkError(err)
 
-	if len(accounts) == 0 && request.Username == "shiori" && request.Password == "gopher" {
+	defaultUser, _ := os.LookupEnv("SHIORI_DEFAULT_USER")
+	defaultPass, _ := os.LookupEnv("SHIORI_DEFAULT_PASS")
+
+	if(defaultUser == "") {
+		defaultUser = "shiori"
+	}
+	if(defaultPass == ""){
+		defaultPass = "gopher"
+	}
+
+	if len(accounts) == 0 && request.Username == defaultUser && request.Password == defaultPass {
 		genSession(model.Account{
-			Username: "shiori",
+			Username: defaultUser,
 			Owner:    true,
 		}, time.Hour)
 		return
